utils/s2n_array.c

```diff
@@ -51,26 +51,46 @@ static S2N_RESULT s2n_array_enlarge(struct s2n_array *array, uint32_t capacity)
 
 struct s2n_array *s2n_array_new(uint32_t element_size)
 {
-    struct s2n_blob mem = {0};
+    struct s2n_array *array = s2n_array_new_with_capacity(element_size, S2N_INITIAL_ARRAY_SIZE);
+    PTR_ENSURE_REF(array);
+
+    return array;
+}
+
+struct s2n_array *s2n_array_new_with_capacity(uint32_t element_size, uint32_t capacity)
+{
+    DEFER_CLEANUP(struct s2n_blob mem = {0}, s2n_free);
     PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_array)));
 
     struct s2n_array *array = (void *) mem.data;
 
-    *array = (struct s2n_array) {.mem = {0}, .len = 0, .element_size = element_size};
+    PTR_GUARD_RESULT(s2n_array_init_with_capacity(array, element_size, capacity));
 
-    if (s2n_result_is_error(s2n_array_enlarge(array, S2N_INITIAL_ARRAY_SIZE))) {
-        /* Avoid memory leak if allocation fails */
-        PTR_GUARD_POSIX(s2n_free(&mem));
-        return NULL;
-    }
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
```

Comment:
Cleanup here is a little tricky. Cleaning up mem might not be enough after s2n_array_init_with_capacity. What if that method failed with s2n_array_validate, but had already successfully called s2n_array_enlarge?

I think you need a DEFER_CLEANUP on array.

Reply:
Ah right okay. I moved the DEFER_CLEANUP to array.